modules/configuration/src/main/java/org/apache/ignite/internal/configuration/ConfigurationChanger.java

```diff
@@ -263,40 +271,63 @@ public void start() {
     }
 
     /**
-     * Initializes the configuration storage - reads data and sets default values for missing configuration properties.
+     * Persist default values to the storage.
      *
-     * @throws ConfigurationValidationException If configuration validation failed.
-     * @throws ConfigurationChangeException If configuration framework failed to add default values and save them to storage.
+     * <p>Make sure all properties that have their values set in the schema defaults are persisted to the storage.
+     * The goal is to keep all configuration values in onl place
+     * and avoid possible cluster inconsistency when code defaults change in the future releases.
+     *
+     * @return Future that will be completed after changes have been written to the storage.
      */
-    public void initializeDefaults() throws ConfigurationValidationException, ConfigurationChangeException {
-        try {
-            ConfigurationSource defaultsCfgSource = new ConfigurationSource() {
-                /** {@inheritDoc} */
-                @Override
-                public void descend(ConstructableTreeNode node) {
-                    addDefaults((InnerNode) node);
-                }
-            };
-
-            changeInternally(defaultsCfgSource).get(5, TimeUnit.SECONDS);
-        } catch (ExecutionException e) {
-            Throwable cause = e.getCause();
+    public CompletableFuture<Void> persistDefaults() {
+        if (storageRoots == null) {
+            throw new ComponentNotStartedException();
+        }
+        if (defaultChanges == null) {
+            return CompletableFuture.completedFuture(null);
+        }
+        return storage.lastRevision()
+                .thenComposeAsync(this::persistDefaultsInternal, pool)
+                .thenRun(() -> defaultChanges = null)
+                .exceptionally(throwable -> {
+                    Throwable cause = throwable.getCause();
+
+                    if (cause instanceof ConfigurationChangeException) {
+                        throw (ConfigurationChangeException) cause;
+                    } else {
+                        throw new ConfigurationChangeException("Failed to persist defaults", cause);
+                    }
+                });
+    }
 
-            if (cause instanceof ConfigurationValidationException) {
-                throw (ConfigurationValidationException) cause;
+    /**
+     * Internal configuration persist method.
+     */
+    private CompletableFuture<Void> persistDefaultsInternal(long storageRevision) {
```

Comment:
I have doubts in the approach you've chosen. Remember that `//Workaround for distributed configuration.` comment in the code? It's there for a reason. And the reason is - it's a workaround, that must be removed. I believe it must be removed in this issue. Sorry for not having proper TODO with the description in that comment, that's on me. Let me explain my position.

Now, this is going to be a pretty long comment, so bear with me.
1. Can this CAS fail? I see that literally every node tries to execute the operation.So it seems like it can. On the other hand, incoming updates should make the map empty. Can we have a concurrent update of the configuration, that will come from a user? Maybe. You probably meant the upgrade from older to a newer version of Ignite.
Are there plans for `init` operation upon upgrade, or we will just quietly add default, invisible to the user? This is a complicated question.
2. I think there might be a simpler way to do things, it's removing that code with a "workaround" comment. This way, all changers will always explicitly write data, that was previously `null`, to the storage (with some tweaks). There are advantages to such approach:
 - Upon init, we would only need a single configuration update, instead of two.
 - We will reuse already existing code, making the procedure much simpler.
One thing to consider - we may still need to call similar method on each start, but having the update with empty map, that will have a specific property of not incrementing the storage revision if there are no updates, would solve the issue, with much less complexity in code.
The only possible issue here, that I see, is that some validators would get `null` from "old" values of configuration, and that must be handled properly. This is also a little inconvenient, because primitive values would throw `NullPointerException` instead of returning nulls, because you know, primitives can't be nulls. This issue can be eliminated, if we will have both "tree with defaults" and "tree without defaults" in the `storageRoots`.
Otherwise, it seems like a better approach to me.
3. Maybe this is not relevant, but this is, most likely, a misuse of the read lock. It has a very specific purpose of protecting a single field. I guess it's not documented properly. But the idea is - it protected the state of the `storageRoots` field while executing `src.descend(changes);`. Your code doesn't use any external closures, and the only access you have to `storageRoots` is a single volatile read, which would still be correct without the read lock.
Considering that I propose deleting the method entirely, this comment was purely informational.

Overall, I would like to discuss my alternative approach with you. Let's do it, please also ask clarifying questions about what I've said, it may be a little confusing, I know that.